src/crypto/equihash.cpp

```diff
@@ -0,0 +1,287 @@
+// Copyright (c) 2016 Jack Grigg
+// Copyright (c) 2016 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+// Implementation of the Equihash Proof-of-Work algorithm.
+//
+// Reference
+// =========
+// Alex Biryukov and Dmitry Khovratovich
+// Equihash: Asymmetric Proof-of-Work Based on the Generalized Birthday Problem
+// NDSS ’16, 21-24 February 2016, San Diego, CA, USA
+// https://www.internetsociety.org/sites/default/files/blogs-media/equihash-asymmetric-proof-of-work-based-generalized-birthday-problem.pdf
+
+#include "crypto/equihash.h"
+#include "util.h"
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <stdexcept>
+
+void validate_params(int n, int k)
+{
+    if (k>=n) {
+        std::cerr << "n must be larger than k\n";
+        throw invalid_params();
+    }
+    if (n % 8 != 0) {
+        std::cerr << "Parameters must satisfy n = 0 mod 8\n";
+        throw invalid_params();
+    }
+    if ((n/(k+1)) % 8 != 0) {
+        std::cerr << "Parameters must satisfy n/(k+1) = 0 mod 8\n";
+        throw invalid_params();
+    }
+}
```

Comment:
``k=0`` is a valid input. This might be acceptable if this leads to 0-length solutions and this short-circuits the whole solution process, which may be useful for testing, diagnostic purposes. That's my preferred behavior, so I'll look for tests with ``k==0``. ;-)

Another alternative is to fail here if ``k == 0``.